README.md
# whitechapelR

***A set of functions for tracking the hidden movement of 'Jack' in the board game Letters from Whitechapel by Fantasy Flight Games***

[![Build Status](https://travis-ci.org/bmewing/whitechapelR.svg?branch=master)](https://travis-ci.org/bmewing/whitechapelR) [![Coverage Status](https://img.shields.io/codecov/c/github/bmewing/whitechapelR/master.svg)](https://codecov.io/github/bmewing/whitechapelR?branch=master) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/whitechapelR)](https://CRAN.R-project.org/package=whitechapelR) ![](http://cranlogs.r-pkg.org/badges/whitechapelR)

## Why do I want this?

There are many reasons you may want to install this package.

 * You've played the game [Letters from Whitechapel](https://boardgamegeek.com/boardgame/59959/letters-whitechapel) *not* as Jack and you've realized how incredibly broken the game is for the policemen if Jack is even halfway competent. 
 * You've played the game Letters from Whitechapel as Jack and want to learn how to make your play even more nefarious.
 * You desire advanced policing tools to help narrow down your search and level the playing field against insidious serial killers. 
 * Your thirst for justice leads you to install anything that promises to help get it.
 * You saw my [amazing Shiny app](https://markewing.shinyapps.io/whitechapel/) and wanted the package for yourself.
 * You're a little confused and think this will help you make delicious sliders (sorry, that's White Castle).

## Install it!

```r   
#Install the latest version from GitHub
#devtools::install_github("bmewing/whitechapelR")
```

## Usage

This package contains a number of functions which, when used together, allow you to map out all the possible places where Jack might be and narrow your field of search.

### Initialize a round

Each round you must indicated where the first murder occurred. This is done with the `start_round()` function which takes as it's only argument the space(s) of the murder. The result needs to be peserved.

```r
paths = start_round(64)
```

### Track Jack's movement

As Jack moves, you'll need to monitor where he might possibly be. There are two functions which help with this, `take_a_step` and `take_a_carriage` (which is a wrapper for `take_a_step`). Taking a step requires that we provide the list of all possible paths taken up to this point and which graph Jack is currently using. By default, he uses the `roads` graph but when using a lantern he uses the `alley` graph. These are both included in the package by default.

```r
#if taking a normal step
paths = take_a_step(paths,roads)

#if taking a latern special movement
# paths = take_a_step(paths,alley)
```

Additionally, you may specify via a list of blocked nodes where Jack cannot move due to being blocked by a policeman. Note, policemen cannot block alley movement but this is not hardcoded into the function, so you could possibly mess up your information at this point.

```r
#if policemen are blocking the road between 50 and 31 as well as between 30 and 13
# paths = take_a_step(paths,roads,list(c(50,31),c(30,13)))
```

Taking a carriage can only be over roads and is never blocked by policemen so it is a simple wrapper for taking two steps with the added functionality of following the rule where Jack cannot end his movement on the space where he began it.

```r
# paths = take_a_carriage(paths)
```

### Looking for clues

As a policeman searching for clues, you want to be able to modify possible paths based on the evidence you aquire. This is done with the `inspect_space` function. It requires the list of all possible paths, the space(s) inspected and a boolean indicating what was found. Because of the exhaustive nature of the methods here, not finding anything can still be incredibly informative. 

```r
#if no clue was found
paths = inspect_space(paths,c(50,30),FALSE)
paths = inspect_space(paths,64,TRUE)
```

It should be pointed out that the algorithms assume you never make mistakes. Finding a clue simply eliminates all paths which do not contain that number. If you input a space which Jack could never have visited and indicate you found a clue there, the path set will be reduced to 0 and you'll have no information from which to work. There is no undo function here.

### Ending a round

Once Jack has (irritatingly) successfully made it their hideout, you want to track which spaces are candidates and improve that information over time.  This is done with the `end_round` function. By storing the endpoints of all possible paths in a variable (say `hideouts`), future rounds can use the intersection of these endpoints and previous information to, hopefully, reduce the possible spaces to a manageable number.

```r
# at the end of the first round
hideouts = end_round(paths)

# at the end of subsequent rounds
# hidouts = end_round(paths,hideouts)
```

### Visualizing Jack's movement

The final utility to mention here is the `show_board` funtion which produces an igraph representation of the game board with the appropriately numbered nodes in their identical positions. Nodes are colored based on the proportion of possible paths which flow through them (key areas to focus on) and can also highlight where hideouts might be.

```r
show_board(paths,hideouts,roads,alley,node_locations)
```


NEWS.md
# whitechapelR 0.3.0

* Fixed a rules bug with carriage movement, now it cannot end on the space where it started

# whitechapelR 0.2.2

* Fixed a bug with the plotting function when passing in hideouts

# whitechapelR 0.2.1

* Added examples to every function.

# whitechapelR 0.2.0

* Initial release to CRAN


tests/testthat.R
library(testthat)
library(whitechapelR)

test_check("whitechapelR")


tests/testthat/test_whitechapel.R
context("Start Round")

test_that("returns list",{
  expect_type(start_round(84),"list")
  expect_type(start_round(c(84,117)),"list")
  expect_identical(start_round(84),list(84))
  expect_identical(start_round(c(84,117)),list(84, 117))
})

context("Take a step")
data(roads)
data(alley)
data(node_locations)

test_that("steps are correct",{
  expect_identical(take_a_step(list(84),roads),                          list(c(84, 51)
                                                                             ,c(84, 65)
                                                                             ,c(84, 66)
                                                                             ,c(84, 67)
                                                                             ,c(84, 85)
                                                                             ,c(84, 86)
                                                                             ,c(84, 99)
                                                                             ,c(84, 100)))
  expect_identical(take_a_step(list(1),roads),                           list(c(1, 2)
                                                                             ,c(1, 6)
                                                                             ,c(1, 7)
                                                                             ,c(1, 8)
                                                                             ,c(1, 9)
                                                                             ,c(1, 24)
                                                                             ,c(1, 26)
                                                                             ,c(1, 28)))
  expect_identical(take_a_step(list(1),alley),                           list(c(1, 7)
                                                                             ,c(1, 26)))
  expect_identical(take_a_carriage(list(23)),                            list(c(23, 21, 20)
                                                                             ,c(23, 21, 40)
                                                                             ,c(23, 21, 41)
                                                                             ,c(23, 21, 42)
                                                                             ,c(23, 22, 42)
                                                                             ,c(23, 22, 77)
                                                                             ,c(23, 77, 22)
                                                                             ,c(23, 77, 75)
                                                                             ,c(23, 77, 76)
                                                                             ,c(23, 77, 90)
                                                                             ,c(23, 77, 91)
                                                                             ,c(23, 77, 92)
                                                                             ,c(23, 77, 93)
                                                                             ,c(23, 77, 94)))
  expect_identical(take_a_step(list(84),roads,list(c(84,51),c(84,100))), list(c(84, 65)
                                                                             ,c(84, 66)
                                                                             ,c(84, 67)
                                                                             ,c(84, 85)
                                                                             ,c(84, 86)
                                                                             ,c(84, 99)))
})

context("Inspecting")

test_that("Inspections are correct",{
  expect_identical(inspect_space(take_a_step(list(1),roads)
                                ,9
                                ,TRUE),list(c(1,9)))
  expect_identical(inspect_space(take_a_step(list(1),alley)
                                ,7
                                ,FALSE),list(c(1,26)))
})

context("End of Round")

test_that("Hideouts correct",{
  expect_identical(end_round(take_a_step(list(1),alley)),c(7,26))
  expect_identical(end_round(take_a_step(list(1),alley),c(26,45)),c(26))
})

context("Visuals")

test_that("Expect plot generates without errors",{
  expect_null(show_board(list(69),c(45,46),roads,alley,node_locations))
  expect_null(show_board(list(69),NULL,roads,alley,node_locations))
})


R/data.R
#' @title Undirected edge pairing of roads connecting nodes
#' @description Data used to establish possible connections used by Jack between nodes
#' @format A data frame with 767 rows and 2 variables
#' \describe{
#'   \item{x}{The smaller integer of the pair of vertices}
#'   \item{y}{The larger integer of the pair of vertices}
#' }
"roads"

#' @title Undirected edge pairing of alley connecting nodes
#' @description Data used to establish possible connections used by Jack between nodes via alleyways
#' @format A data frame with 452 rows and 2 variables
#' \describe{
#'   \item{x}{The smaller integer of the pair of vertices}
#'   \item{y}{The larger integer of the pair of vertices}
#' }
"alley"

#' @title x,y coordinates of node points from the game board
#' @description Data used to place nodes in graphical output according to their relative positions on the game board
#' @format A data frame with 195 rows and 4 variables
#' \describe{
#'   \item{id}{An artifact of the computer vision process used to obtain coordinates}
#'   \item{x}{The number of pixels from the left edge of the board to the center of the node}
#'   \item{y}{The number of pixels from the top edge of the board to the center of the node}
#'   \item{name}{The integer assigned to the node on the game board}
#' }
"node_locations"


R/end_round.R
#' @export

end_round = function(paths,hideouts=NULL){
  #' @title Manage list of possible hideouts
  #'
  #' @description Create or update a list of possible hideouts based on final positions from the list of possible paths traveled.
  #'
  #' @param paths list of all possible paths already traveled
  #' @param hideouts optional vector of possible hideouts from previous rounds. Not used in round 1, only rounds 2 and 3
  #'
  #' @return list of all possible hideouts
  #' @examples
  #' possibilities = start_round(64)
  #' possibilities = take_a_step(possibilities,roads)
  #' possibilities = take_a_step(possibilities,roads,blocked=list(c(63,82),c(63,65)))
  #' possibilities = inspect_space(possibilities,space = c(29,30), clue = FALSE)
  #' possibilities = inspect_space(possibilities,space = 49, clue = TRUE)
  #' hideouts = end_round(possibilities,hideouts=NULL)
  #' possibilities = start_round(67)
  #' possibilities = take_a_step(possibilities,roads)
  #' hideouts = end_round(possibilities,hideouts=hideouts)

  possible_hideouts = vapply(paths, function(x){
    x[length(x)]
  }, numeric(1))
  possible_hideouts = unique(possible_hideouts)
  if(is.null(hideouts)) return(sort(possible_hideouts))
  hideouts = intersect(hideouts,possible_hideouts)
  return(sort(hideouts))
}


R/show_board.R
#' @export

show_board = function(paths=NULL,hideouts=NULL,roads,alley,node_locations){
  #' @title Display game board representation
  #'
  #' @description Show a graph representation of the game board with nodes placed in the appropriate relative spot, colored by the number of paths which include a particular node. Possible hideouts are marked with blue squares.
  #'
  #' @param paths optional list of all possible paths already traveled
  #' @param hideouts optional vector of possible hideouts from previous rounds.
  #' @param roads data.frame of non-directional edge pairs for the road graph
  #' @param alley data.frame of non-directional edge pairs for the alley graph
  #' @param node_locations data.frame of where nodes should be placed in the graph
  #'
  #' @return plotted igraph object
  #'
  #' @details
  #' roads, alley and node_locations are all bundled with the package (e.g. \code{data(roads)}).
  #' Solid lines in the graph represent road connections between nodes. Dashed lines represent alley way connections.
  #'
  #' @examples
  #' possibilities = start_round(64)
  #' possibilities = take_a_step(possibilities,roads)
  #' possibilities = take_a_step(possibilities,roads,blocked=list(c(63,82),c(63,65)))
  #' possibilities = take_a_step(possibilities,alley)
  #' show_board(paths=possibilities,hideouts=NULL,roads,alley,node_locations)

  r = roads
  a = alley
  r$lty = 1
  a$lty = 3
  r$weight = 1
  a$weight = 1
  l = as.matrix(node_locations[order(node_locations$name),c("x","y")])
  l[,2] = l[,2]*-1
  v = data.frame(name = 1:195
                ,cex=0.7
                ,color="white"
                ,shape="circle"
                ,stringsAsFactors = FALSE)

  if(!is.null(paths)){
    colors = c("#FFD9D9","#F5C5C5","#ECB1B1","#E39D9D","#D98A8A","#D07676","#C76262","#BE4E4E","#B43B3B","#AB2727","#A21313","#990000")
    tbl = table(unlist(paths))
    to_replace = colors[round(tbl/max(tbl)*10,0)]
    v$color[as.numeric(names(tbl))] = to_replace
  }

  if(!is.null(hideouts)){
    v$shape[hideouts] = "square"
    v$color[hideouts] = "sky blue"
  }

  g = igraph::graph_from_data_frame(rbind(r,a),directed = FALSE,vertices = v)

  par(mai=c(0,0,0,0))
  plot(g,layout=l,vertex.size = 5)
}


R/inspect_space.R
#' @export

inspect_space = function(paths,space,clue){
  #' @title Update paths based on inspections
  #'
  #' @description Updated the list of possible paths based on the results of police investigation
  #'
  #' @param paths list of all possible paths already traveled
  #' @param space vector of integers of the spaces inspected
  #' @param clue single logical value indicating if evidence of Jack was found
  #'
  #' @return list of all possible paths traveled by Jack
  #' @examples
  #' possibilities = start_round(64)
  #' possibilities = take_a_step(possibilities,roads)
  #' possibilities = take_a_step(possibilities,roads,blocked=list(c(63,82),c(63,65)))
  #' possibilities = inspect_space(possibilities,space = c(29,30), clue = FALSE)
  #' possibilities = inspect_space(possibilities,space = 49, clue = TRUE)

  if(clue){
    paths = lapply(paths,found_clue,space=space)
  } else {
    paths = lapply(paths,found_nothing,space=space)
  }
  return(plyr::compact(paths))
}

found_clue = function(path,space){
  if(!any(path %in% space)) return(NULL)
  return(path)
}

found_nothing = function(path,space){
  if(any(path %in% space)) return(NULL)
  return(path)
}


R/take_a_step.R
globalVariables(c("roads"))

take_a_step = function(paths,roads,blocked = NULL){
  #' @title Track one movement
  #'
  #' @description Track one step of unknown movement by Jack, either on roads or through alleyways
  #'
  #' @param paths list of all possible paths already traveled
  #' @param roads data.frame of non-directional edge pairs for either the road graph or the alley graph
  #' @param blocked list of node pairs which cannot be traversed because a police officer blocks it (should not be used for special movement)
  #'
  #' @return list of all possible paths traveled by Jack
  #'
  #' @details
  #' The non-directional edge pairs are available via \code{data(roads)} or \code{data(alley)}
  #' This function does not account for the rule that Jack cannot travel through a road occupied by a police officer.
  #'
  #' @examples
  #' possibilities = start_round(64)
  #' possibilities = take_a_step(possibilities,roads)
  #' possibilities = take_a_step(possibilities,roads,blocked=list(c(63,82),c(63,65)))
  #' possibilities = take_a_step(possibilities,alley)

  paths = lapply(paths,gen_possibilities,roads=roads,blocked=blocked)
  paths = unlist(paths,recursive = FALSE)

  return(paths)
}

take_a_carriage = function(paths){
  #' @title Track carriage movement
  #'
  #' @description Track two steps of unknown movement by Jack, on roads
  #'
  #' @param paths list of all possible paths already traveled
  #'
  #' @return list of all possible paths traveled by Jack
  #' @examples
  #' possibilities = start_round(64)
  #' possibilities = take_a_carriage(possibilities)

  data(roads,envir = environment())
  paths = take_a_step(paths,roads)
  paths = take_a_step(paths,roads)
  paths = Filter(function(x){rx = rev(x); rx[1] != rx[3]}, paths)
  return(paths)
}

trim_possibilities = function(paths,node){
  #' @title Trim possible paths
  #'
  #' @description Remove known impossible end points for Jack, typically as a result of having found, but not arrested Jack.
  #'
  #' @param paths list of all possible paths already traveled
  #' @param node vector of length 1 or 2 which specifies blocked nodes due to the presence of a policeman
  #'
  #' @return list of trimmed possible paths traveled by Jack
  #' @examples
  #' possibilities = start_round(64)
  #' possibilities = take_a_carriage(possibilities)
  #' possibilities = trim_possibilities(possibilities,82)
  #' possibilities = trim_possibilities(possibilities,c(66,67))
  keep = unlist(lapply(paths,function(x){!identical(sort(rev(x)[seq_along(node)]),sort(node))}))
  return(paths[keep])
}

gen_possibilities = function(path,roads,blocked = NULL){
  start = rev(path)[1]
  p = unique(unlist(roads[roads$x == start | roads$y == start,]))
  p = p[p != start]
  full_paths = lapply(p,add_possibilities,path=path)
  if(!is.null(blocked)){
    r = lapply(blocked,trim_possibilities,paths=full_paths)
    full_paths = Reduce(intersect,r,init=r[[1]])
  }
  return(full_paths)
}

add_possibilities = function(p,path){
  return(c(path,p))
}
